File: My-Homeworks-ExamPreparations-And-Excersizes/JavaScript Fundamentals/08. Functions/Homework/P06. First larger than neighbours/P06. First larger than neighbours.js
/**
 * Created by Admin on 29.6.2016 г..
 */
/**
 * First larger than neighbours

 Description

 Write a method that returns the index of the first element in array that is larger than its neighbours, or  -1 , if there is no such element.

 Input
 •On the first line you will receive the number N - the size of the array
 •On the second line you will receive N numbers separated by spaces - the array

 Output
 •Print the index of the first element that is larger than its neighbours or  -1  if none are

 Constraints
 •1 <= N <= 1024
 •Time limit: 0.2s
 •Memory limit: 16MB

 Sample tests


 Input

 Output


 6
 -26 -25 -28 31 2 27 1

 Submission
 * @param params
 * @returns {*}
 */

// [ '6', '-26 -25 -28 31 2 27' ]
// var input = [ '6', '-26 -25 -28 31 2 27' ];
// solve(input);

function solve(args) {

    var N = parseInt(args[0]);
    var inputArr = ("" + args[1]);
    inputArr = inputArr.split(/\s+/g);
    inputArr = inputArr.map(function (item) {
        return parseInt(item, 10);
    });

    console.log(indexOfLargerThanNeighbour(inputArr));

    function indexOfLargerThanNeighbour(numbers) {
        var ixOfFirstLargerThanNeighbours = -1;

        for(i = 0; i < numbers.length; i = i + 1){

            var isLargerThanNeighburs = false;
            if(i == 0 && i !== numbers.length - 1){
                // isLargerThanNeighburs = (numbers[i] > numbers[i + 1]);
            }else if(i == numbers.length - 1 && i !== 0){
                // isLargerThanNeighburs = (numbers[i] > numbers[i - 1]);
            }
            else{
                isLargerThanNeighburs = (numbers[i] > numbers[i - 1] && numbers[i] > numbers[i + 1]);
            }

            if (isLargerThanNeighburs){
                ixOfFirstLargerThanNeighbours = i;
                break;
            }
        }

        return ixOfFirstLargerThanNeighbours;
    }
}

// function solve(params) {
//     var arr = ((params+'').split('\n'))[1].split(' ');
//
//     for(var i = 0; i < arr.length; i+=1){
//         if (+arr[i] > +arr[i-1] && +arr[i] > +arr[i+1]) {
//             return i;
//         }
//     }
//     return -1;
// }

File: My-Homeworks-ExamPreparations-And-Excersizes/CSharp-Advanced/03. Methods/Homework/P06. First larger than neighbours/P06. First larger than neighbours.cs
/*
First larger than neighbours
Description

Write a method that returns the index of the first element in array that is larger than its neighbours, or -1, if there is no such element.

Input

On the first line you will receive the number N - the size of the array
On the second line you will receive N numbers sepated by spaces - the array
Output

Print the index of the first element that is larger than its neighbours or -1 if none are
Constraints

1 <= N <= 1024
Time limit: 0.1s
Memory limit: 16MB
Sample tests

Input	Output
6
-26 -25 -28 31 2 27	
1
Submission
 */
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace P06.First_larger_than_neighbours
{
    class Program
    {
        static void Main(string[] args)
        {
            int N = int.Parse(Console.ReadLine());
            List<int> nums = ReadInLineArray();
            Console.WriteLine(FirstLargerThenNeighbrs(nums));

        }

        static List<int> ReadInLineArray()
        {
            string inLine = Console.ReadLine();
            //string inLine = "-26 -25 -28 31 2 27	2";

            char[] delimiters = new char[] { ' ', ',' };
            List<int> nums = inLine.Split(delimiters, StringSplitOptions.RemoveEmptyEntries).Select(d => int.Parse(d)).ToList();

            return nums;
        }

        static int CountLargerThenNeighbrs(List<int> numbers)
        {
            int countLargerThenNeighbrc = 0;
            bool isLargerThenNeighbrs = false;

            int firstIx = 0;
            int lastIx = numbers.Count - 1;

            //Array from one element
            if (firstIx == lastIx)
            {
                return countLargerThenNeighbrc;
            }
            //-26 -25 -28 31 2 27	
            for (int i = 0; i < lastIx + 1; i++)
            {
                if (i == firstIx)
                {
                    //isLargerThenNeighbrs = (numbers[i] > numbers[i + 1]);
                }
                else if (i == lastIx)
                {
                    //isLargerThenNeighbrs = (numbers[i] > numbers[i - 1]);
                }
                else
                {
                    isLargerThenNeighbrs = ((numbers[i - 1] < numbers[i]) && (numbers[i] > numbers[i + 1]));

                }

                if (isLargerThenNeighbrs)
                {
                    countLargerThenNeighbrc++;
                    isLargerThenNeighbrs = false;
                }
            }

            return countLargerThenNeighbrc;
        }

        static int FirstLargerThenNeighbrs(List<int> numbers)
        {
            int firstLargerThenNeighbrc = 0;
            bool isLargerThenNeighbrs = false;

            int firstIx = 0;
            int lastIx = numbers.Count - 1;

            //Array from one element
            if (firstIx == lastIx)
            {
                return firstLargerThenNeighbrc;
            }
            //-26 -25 -28 31 2 27	
            for (int i = 0; i < lastIx + 1; i++)
            {
                if (i == firstIx)
                {
                    //isLargerThenNeighbrs = (numbers[i] > numbers[i + 1]);
                }
                else if (i == lastIx)
                {
                    //isLargerThenNeighbrs = (numbers[i] > numbers[i - 1]);
                }
                else
                {
                    isLargerThenNeighbrs = ((numbers[i - 1] < numbers[i]) && (numbers[i] > numbers[i + 1]));

                }

                if (isLargerThenNeighbrs)
                {
                    firstLargerThenNeighbrc = i;                    
                    break;
                }
            }

            return firstLargerThenNeighbrc;
        }
    }
}
